services/web/server/tests/unit/test_security_access_model.py

```diff
@@ -0,0 +1,283 @@
+    # ANONYMOUS <--- USER <--- TESTER
+
+    R = UserRole
+
+    OPERATION = "studies.templates.read"
+    assert await access_model.can(R.ANONYMOUS, OPERATION)
+    assert await access_model.can(R.USER, OPERATION)
+    assert await access_model.can(R.TESTER, OPERATION)
+
+    OPERATION = "study.node.create"
+    assert not await access_model.can(R.ANONYMOUS, OPERATION)
+    assert await access_model.can(R.USER, OPERATION)
+    assert await access_model.can(R.TESTER, OPERATION)
+
+    OPERATION = "study.nodestree.uuid.read"
+    assert not await access_model.can(R.ANONYMOUS, OPERATION)
+    assert not await access_model.can(R.USER, OPERATION)
+    assert await access_model.can(R.TESTER, OPERATION)
+
+    OPERATION = "study.amazing.action"
+    assert not await access_model.can(R.ANONYMOUS, OPERATION)
+    assert not await access_model.can(R.USER, OPERATION)
+    assert not await access_model.can(R.TESTER, OPERATION)
+
+
+
+async def test_checked_permissions(access_model):
+    R = UserRole # alias
+    MOCKPATH = 'data/fake-template-projects.isan.old.json'
+
+    with resources.stream(MOCKPATH) as fh:
+        data = json.load(fh)
+
+    current ={}
+    for prj in data:
+        if prj["uuid"] == "template-uuid-1234-a1a7-f7d4f3a8f26b":
+            current = prj
+            break
+
+    assert current, "Did '%s' changed??" % MOCKPATH
+
+    # updates both allowed and not allowed fields
+    candidate = copy.deepcopy(current)
+    candidate['workbench']['template-uuid-409d-998c-c1f04de67f8b']["inputs"]["Kr"] = 66 # ReadOnly!
+    candidate['workbench']['template-uuid-409d-998c-c1f04de67f8b']["inputs"]["Na"] = 66 # ReadWrite
+
+    assert not await access_model.can(
+        R.ANONYMOUS,
+        "study.pipeline.node.inputs.update",
+        context={'current': current, 'candidate': candidate}
+    )
+
+    # updates allowed fields
+    candidate = copy.deepcopy(current)
+    candidate['workbench']['template-uuid-409d-998c-c1f04de67f8b']["inputs"]["Na"] = 66 # ReadWrite
+
+    assert await access_model.can(
+        R.ANONYMOUS,
+        "study.pipeline.node.inputs.update",
+        context={'current': current, 'candidate': candidate}
+    )
+
+    # udpates not permitted fields
+    candidate = copy.deepcopy(current)
+    candidate['description'] = 'not allowed to write here'
+    assert not await access_model.can(
+        R.ANONYMOUS,
+        "study.pipeline.node.inputs.update",
+        context={'current': current, 'candidate': candidate}
+    )
+
+
+async def test_async_checked_permissions(access_model):
+    R = UserRole # alias
+
+    # add checked permissions
+    async def async_callback(context) -> bool:
+        return context['response']
+
+    access_model.roles[R.TESTER].check['study.edge.edit'] = async_callback
+
+    assert not await access_model.can(
+        R.TESTER,
+        "study.edge.edit",
+        context={'response':False}
+    )
+
+    assert await access_model.can(
+        R.TESTER,
+        "study.edge.edit",
```

Comment:
what means .edge. in that context?

Reply:
this is a dummy taken from the front end. It refers to the edge of a graph but we are not using this action in the backend